src/librustc_codegen_ssa/base.rs

```diff
@@ -606,20 +607,62 @@ pub fn codegen_crate<B: ExtraBackendMethods>(
         codegen_units
     };
 
-    let mut total_codegen_time = Duration::new(0, 0);
+    let total_codegen_time = Lock::new(Duration::new(0, 0));
 
-    for cgu in codegen_units.into_iter() {
+    let cgu_reuse: Vec<_> = tcx.sess.time("find cgu reuse", || {
+        codegen_units.iter().map(|cgu| determine_cgu_reuse(tcx, &cgu)).collect()
+    });
+
+    let mut cgus: FxHashMap<usize, _> = if cfg!(parallel_compiler) {
+        tcx.sess.time("compile first CGUs", || {
+            // Try to find one CGU to compile per thread.
+            let cgus: Vec<_> = cgu_reuse
+                .iter()
+                .enumerate()
+                .filter(|&(_, reuse)| reuse == &CguReuse::No)
+                .take(tcx.sess.threads())
+                .collect();
+
+            // Compile the found CGUs in parallel.
+            par_iter(cgus)
```

Comment:
Rayon will take care of coordinating with the job server, right?

Reply:
The worker thread might use the implicit token for something in the background, if so I might have to move this into the loop.

Reply:
The worker thread will indeed use the implicit token, so this code has been moved into the loop.